File: 0449-Serialize and Deserialize BST/0449-Serialize and Deserialize BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (root == nullptr)
        {
            return "";
        }
        
        vector<TreeNode*> Q;
        Q.push_back(root);
        for (int i = 0; i < Q.size(); ++i)
        {
            TreeNode* node = Q[i];
            if (node == nullptr)
            {
                continue;
            }
            Q.push_back(node->left);
            Q.push_back(node->right);
        }
        
        while (Q.back() == nullptr)
        {
            Q.pop_back();
        }
        
        string result;
        for (TreeNode* node : Q)
        {
            if (node == nullptr)
            {
                result += ",null";
            }
            else if (result.size())
            {
                result += "," + to_string(node->val);
            }
            else
            {
                result += to_string(node->val);
            }
        }
        
        return result;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.size() == 0)
        {
            return nullptr;
        }
        stringstream ss(data);
        queue<TreeNode*> Q;
        string s;
        TreeNode* root = nullptr;
        bool left = true;
        while (getline(ss, s, ','))
        {
            if (root == nullptr)
            {
                root = new TreeNode(stoi(s));
                Q.push(root);
            }
            else
            {
                if (s != "null")
                {
                    TreeNode* node = new TreeNode(stoi(s));
                    if (left)
                    {
                        Q.front()->left = node;
                    }
                    else
                    {
                        Q.front()->right = node;
                    }
                    Q.push(node);
                }
                
                if (!left)
                {
                    Q.pop();
                }
                
                left = !left;
            }
        }
        
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: 0401-0500/0449-Serialize and Deserialize BST/0449-Serialize and Deserialize BST.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

import collections
class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if root is None:
            return '[]'
        nodes = [root]
        i = 0
        while i < len(nodes):
            if nodes[i]:
                nodes.append(nodes[i].left)
                nodes.append(nodes[i].right)
            i += 1
        
        while nodes[-1] is None:
            nodes.pop()
            
        return '[' + ','.join(str(node.val) if node else 'null' for node in nodes) +']'
            

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if data == '[]':
            return None
        
        strs = data[1:-1].split(',')
        root = TreeNode(int(strs[0]))
        Q = collections.deque([root])
        leftChild = True
        for s in strs[1:]:
            if s != 'null':
                node = TreeNode(int(s))
                if leftChild:
                    Q[0].left = node
                else:
                    Q[0].right = node
                Q.append(node)
            leftChild =  not leftChild
            if leftChild:
                Q.popleft()
        return root    
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
